cmd/store.go

```diff
@@ -31,64 +35,132 @@ func (s *store) Get(IP string) (string, error) {
 	return "", fmt.Errorf("Unable to find role for IP %s", IP)
 }
 
-// OnAdd is called when a pod is added.
-func (s *store) OnAdd(obj interface{}) {
-	pod, ok := obj.(*api.Pod)
-	if !ok {
-		log.Errorf("Expected Pod but OnAdd handler received %+v", obj)
-		return
+func (s *store) AddRoleToIP(pod *api.Pod, role string) {
+	s.mutex.Lock()
+	s.rolesByIP[pod.Status.PodIP] = role
+	s.mutex.Unlock()
+}
+
+func (s *store) AddNamespaceToIP(pod *api.Pod) {
+	s.mutex.Lock()
+	s.namespaceByIP[pod.Status.PodIP] = pod.GetNamespace()
+	s.mutex.Unlock()
+}
+
+func (s *store) DeleteIP(ip string) {
+	s.mutex.Lock()
+	delete(s.rolesByIP, ip)
+	delete(s.namespaceByIP, ip)
+	s.mutex.Unlock()
+}
+
+// AddRoleToNamespace takes a role name and adds it to our internal state
+func (s *store) AddRoleToNamespace(namespace string, role string) {
+	// Make sure to add the full ARN of roles to ensure string matching works
+	roleARN := s.iam.roleARN(role)
+
+	ar := s.rolesByNamespace[namespace]
+	if ar == nil {
+		ar = []string{}
 	}
 
-	if pod.Status.PodIP != "" {
-		if role, ok := pod.Annotations[s.iamRoleKey]; ok {
-			s.mutex.Lock()
-			s.rolesByIP[pod.Status.PodIP] = role
-			s.mutex.Unlock()
+	// this is a tiny bit troubling, we could go with a the rolesByNamespace
+	// being a map[string]map[string]bool so that deduplication isn't
+	// ever a problem .. but for now...
+	c := true
+	for i := range ar {
+		if ar[i] == roleARN {
+			c = false
+			break
 		}
 	}
+	if c {
+		ar = append(ar, roleARN)
+	}
+	s.mutex.Lock()
+	s.rolesByNamespace[namespace] = ar
+	s.mutex.Unlock()
```

Comment:
I think it's really just a style thing. As far as I can tell for simple things like this, explicit lock/unlock code will be ever so slightly more performant but in reality it's probably negligible. Happy to change this if people think it looks better, otherwise I'll stick with what was already there.